k2/python/csrc/torch/index_add.cu

```diff
@@ -52,68 +57,149 @@ void IndexAdd(torch::Tensor index, torch::Tensor value, torch::Tensor *in_out) {
+    return;
+  }
 
-        K2_DCHECK_GE(in_out_i, 0);
-        K2_DCHECK_LT(in_out_i, in_out_numel);
+  // for non-contiguous tensors
+  int64_t index_stride = index.strides()[0];
+  int64_t value_stride = value.strides()[0];
+  int64_t in_out_stride = in_out->strides()[0];
 
-        // for host thread, we assume single thread at present
-        in_out_data[in_out_i] += value_data[i];
-      }
-    } else {
-      // for cuda
-      auto lambda_contiguous = [=] __device__(int32_t i) -> void {
-        int32_t in_out_i = index_data[i];
+  K2_EVAL(
+      context, index_numel, lambda_noncontiguous, (int32_t i)->void {
+        int32_t in_out_i = index_data[i * index_stride];
         if (in_out_i == -1) return;
 
         K2_DCHECK_GE(in_out_i, 0);
         K2_DCHECK_LT(in_out_i, in_out_numel);
 
-        atomicAdd(in_out_data + in_out_i, value_data[i]);
-      };
-      EvalDevice(context, index_numel, lambda_contiguous);
-    }
+#ifdef __CUDA_ARCH__
+        atomicAdd(in_out_data + in_out_i * in_out_stride,
+                  value_data[i * value_stride]);
+#else
+      // for host thread, we assume single thread at present
+      in_out_data[in_out_i * in_out_stride] += value_data[i * value_stride];
+#endif
+      });
+}
+
+static void IndexAdd2D(torch::Tensor index, torch::Tensor value,
+                       torch::Tensor *in_out) {
+  NVTX_RANGE(K2_FUNC);
+  K2_CHECK_EQ(index.scalar_type(), ToScalarType<int32_t>::value);
+  K2_CHECK_EQ(index.dim(), 1);
+
+  K2_CHECK_EQ(value.scalar_type(), ToScalarType<float>::value)
+      << "index_add is supposed to be used in back propagation, "
+         "which means the value type should be of float";
+  K2_CHECK_EQ(value.dim(), 2);
+
+  K2_CHECK_EQ(in_out->scalar_type(), ToScalarType<float>::value);
+  K2_CHECK_EQ(in_out->dim(), 2);
+  K2_CHECK_EQ(in_out->sizes()[1], value.sizes()[1]);
+
+  const int32_t *index_data = index.data_ptr<int32_t>();
+  const float *value_data = value.data_ptr<float>();
+  float *in_out_data = in_out->data_ptr<float>();
+
+  ContextPtr context;
+  if (index.device().type() == torch::kCPU) {
+    context = GetCpuContext();
+  } else {
+    K2_CHECK(index.is_cuda());
+    context = GetCudaContext(index.device().index());
+  }
+
+  int32_t index_numel = index.numel();
+  int32_t in_out_num_rows = static_cast<int32_t>(in_out->sizes()[0]);
+  int32_t in_out_num_cols = static_cast<int32_t>(in_out->sizes()[1]);
+
+  if (index.is_contiguous() && value.is_contiguous() &&
+      in_out->is_contiguous()) {
+    K2_EVAL(
+        context, index_numel, lambda_contiguous, (int32_t i)->void {
+          int32_t in_out_i = index_data[i];
+          if (in_out_i == -1) return;
+
+          K2_DCHECK_GE(in_out_i, 0);
+          K2_DCHECK_LT(in_out_i, in_out_num_rows);
+          float *cur_in_out_data = in_out_data + in_out_i * in_out_num_cols;
+          const float *cur_value_data = value_data + i * in_out_num_cols;
+          for (int32_t j = 0; j != in_out_num_cols; ++j) {
```

Comment:
seems there's a loop in kernel, can't we add `j` in the parameter of lambda and use `Eval2`? BTW, I think we need also add a `TODO` here as we definitely need to find a way to avoid `atomicAdd` in future.

Reply:
The number of loops is determined by the number of entries in the arc map, which is a 1-D tensor, so `Eval` is used here. 

I didn't find a way to avoid the loop here.

Reply:
we can just do like below
```
 K2_EVAL2(
        context, index_numel, *in_out_num_cols*,  lambda_contiguous, (int32_t i, *int32_t j*)->void {
          int32_t in_out_i = index_data[i];
          if (in_out_i == -1) return;

          K2_DCHECK_GE(in_out_i, 0);
          K2_DCHECK_LT(in_out_i, in_out_num_rows);
          float *cur_in_out_data = in_out_data + in_out_i * in_out_num_cols;
          const float *cur_value_data = value_data + i * in_out_num_cols;
#ifdef __CUDA_ARCH__
            atomicAdd(cur_in_out_data + j, cur_value_data[j]);
#else
            // for host thread, we assume single thread at present
            cur_in_out_data[j] += cur_value_data[j];
#endif
```

Reply:
We have a __host__ version of atomicAdd so you don't need the ifdef.

Reply:
Fixed.

There was no `__host__ `version of atomicAdd.  A host version is added in this pullrequest.